### PR TITLE
Refactor chart API endpoints

### PR DESCRIPTION
**Pull request by steven-zou**
**Refactor chart API endpoints**

- add /chartrepo after /api
- add /chartrepo before /index.yaml,/:repo/index.yaml and /:repo/charts/:filename
- add debug messages to the potential debug points
- add more supporting mime types "multipart/form-data", "application/octet-stream" to the API endpoints

**Update the related build scripts to package the chart repo server**

- add env file template for chart repo server in make/common/config/chartserver
- update the Makefiles to support build chart repo server
- add docker file and related build scripts for upstream chart server - chartmuseum
- update prepare to support generating chart server related configs
- add docker compose file for the chart server
- add build/install command options to install with/without chart repo server
- update install.sh to support chart repo server installation

**Fix conflicts in Makefiles and prepare script files with upstream**

**Keep the chart server related configurations in adminserver**

- append chart server related config options to the supporting list of adminserver
- provide chart server related config access method in the API layer
- update prepare script and ui env template file to enable cache driver config for chart server API
- append flag info in the systeminfo API to indicate if chart server is deployed with Harbor
- refactor the response rewriting logic to return structural error object
- add API init method to initializing objects required in API handlers
- chage owner of the storage folder
- update offline/online package scripts in Harbor-Util.robot

**NOTES: To pass CI cases for the go files in this PR, the updates of Makefiles/prepare should be counted in**

### CODE REVIEW

File: src/chartserver/reverse_proxy.go [hunk 1]
```diff
@@ -73,25 +74,39 @@ func director(target *url.URL, cred *Credential, req *http.Request) {
...
+	//Detect the 401 code, if it is,overwrite it to 500.
+	//We also re-write the error content to structural error object
+	errorObj := make(map[string]string)
 	if res.StatusCode == http.StatusUnauthorized {
-		errorObj := make(map[string]string)
 		errorObj["error"] = "operation request from unauthentic source is rejected"
-		content, err := json.Marshal(errorObj)
+		res.StatusCode = http.StatusInternalServerError
+	} else {
+		//Exttact the error and wrap it into the error object
```
Comment (reasonerjt) [inline src/chartserver/reverse_proxy.go]: Exttact -> Extract

File: src/chartserver/reverse_proxy.go [hunk 2]
```diff
@@ -73,25 +74,39 @@ func director(target *url.URL, cred *Credential, req *http.Request) {
...
+	if res.StatusCode >= http.StatusOK && res.StatusCode <= http.StatusTemporaryRedirect {
+		return nil
+	}
+
+	//Detect the 401 code, if it is,overwrite it to 500.
+	//We also re-write the error content to structural error object
+	errorObj := make(map[string]string)
 	if res.StatusCode == http.StatusUnauthorized {
-		errorObj := make(map[string]string)
 		errorObj["error"] = "operation request from unauthentic source is rejected"
```
Comment (reasonerjt) [inline src/chartserver/reverse_proxy.go]: unauthentic -> unauthorized.

File: src/ui/api/chart_repository.go [hunk 3]
```diff
@@ -164,6 +175,41 @@ func (cra *ChartRepositoryAPI) UploadChartProvFile() {
 	chartController.GetManipulationHandler().UploadProvenanceFile(cra.Ctx.ResponseWriter, cra.Ctx.Request)
 }
 
+//Rewrite the incoming URL with the right backend URL pattern
```
Comment (reasonerjt) [inline src/ui/api/chart_repository.go]: UT?

File: make/common/templates/adminserver/env [hunk 4]
```diff
@@ -61,5 +61,7 @@ REGISTRY_STORAGE_PROVIDER_NAME=$storage_provider_name
 READ_ONLY=false
 SKIP_RELOAD_ENV_PATTERN=$skip_reload_env_pattern
 RELOAD_KEY=$reload_key
+CHART_SERVICE_URL=$chart_service_url
 LDAP_GROUP_ADMIN_DN=$ldap_group_admin_dn
-REGISTRY_CONTROLLER_URL=$registry_controller_url
\ No newline at end of file
+REGISTRY_CONTROLLER_URL=$registry_controller_url
+WITH_CHARTMUSEUM=$with_chartmuseum
```
Comment (reasonerjt) [inline make/common/templates/adminserver/env]: Why call it CHART_SERVICE_URL?   Not CHART_MUSEUM_URL?  If it's not empty string I think we can assume WITH_CHARTMUSEUM is true?

File: src/ui/config/config.go [hunk 5]
```diff
@@ -540,3 +541,31 @@ func ReadOnly() bool {
 	}
 	return utils.SafeCastBool(cfg[common.ReadOnly])
 }
+
+// WithChartMuseum returns a bool to indicate if chartmuseum is deployed with Harbor.
+func WithChartMuseum() bool {
+	cfg, err := mg.Get()
```
Comment (reasonerjt) [inline src/ui/config/config.go]: It can be implied by ChartMuseumEndpoint?

File: make/common/templates/ui/env [hunk 6]
```diff
@@ -7,3 +7,4 @@ ADMINSERVER_URL=$adminserver_url
 UAA_CA_ROOT=/etc/ui/certificates/uaa_ca.pem
 _REDIS_URL=$redis_url
 SYNC_REGISTRY=false
+CHART_CACHE_DRIVER=$chart_cache_driver
```
Comment (reasonerjt) [inline make/common/templates/ui/env]: How is it used?

File: .travis.yml [hunk 7]
```diff
@@ -28,6 +28,7 @@ env:
   SELF_REGISTRATION: on
   KEY_PATH: /data/secretkey
   REDIS_HOST: localhost
+  BACKEND_CHART_SERVER: http://localhost:9090
```
Comment (reasonerjt) [inline .travis.yml]: How is it used?

File: make/common/templates/adminserver/env (hunk 4 above)
Comment (reasonerjt) [inline make/common/templates/adminserver/env]: WITH_CHARTMUSEUM can be implied by CHART_SERVICE_URL?

File: make/common/templates/ui/env (hunk 6 above)
Comment (reasonerjt) [inline make/common/templates/ui/env]: how is this used?

File: make/docker-compose.chartmuseum.tpl [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  redis:
+    networks:
+      harbor-chartmuseum:
+        aliases:
+          - redis
+  chartmuseum:
+    container_name: chartmuseum
+    image: vmware/chartmuseum-photon:__chartmuseum_version__
+    restart: always
+    cpu_quota: 150000
```
Comment (reasonerjt) [inline make/docker-compose.chartmuseum.tpl]: why? cpu_quota?

File: make/photon/chartserver/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,23 @@
+FROM vmware/photon:1.0
```
Comment (reasonerjt) [inline make/photon/chartserver/Dockerfile]: Should be `photon:1.0`

File: make/prepare [hunk 10]
```diff
@@ -635,5 +656,99 @@ if args.clair_mode:
 if args.ha_mode:
     prepare_ha(rcp, args)
 
+# config chart repository
+if args.chart_mode:
+    chartm_temp_dir = os.path.join(templates_dir, "chartserver")
+    chrtm_config_dir = os.path.join(config_dir, "chartserver")
```
Comment (reasonerjt) [inline make/prepare]: why chartm vs chrtm?

File: src/adminserver/systemcfg/systemcfg.go [hunk 11]
```diff
@@ -220,6 +225,11 @@ var (
 		common.ClairURL:                    "CLAIR_URL",
 		common.NotaryURL:                   "NOTARY_URL",
 		common.DatabaseType:                "DATABASE_TYPE",
+		common.ChartRepoURL:                "CHART_SERVICE_URL",
```
Comment (reasonerjt) [inline src/adminserver/systemcfg/systemcfg.go]: There are different names for the same thing:
ChartRepo, CHART_SERVICE, ChartMuseum ...

File: make/common/templates/ui/env (hunk 6 above)
Comment (steven-zou) [reply make/common/templates/ui/env]: ‘memory' or 'redis'. If it is redis, then try to read _REDIS_URL

File: src/ui/api/chart_repository.go [hunk 12]
```diff
@@ -164,6 +175,41 @@ func (cra *ChartRepositoryAPI) UploadChartProvFile() {
...
+//Remove 'chartrepo' from the endpoints of manipulation API
+//Remove 'chartrepo' from the endpoints of repository services
+func (cra *ChartRepositoryAPI) rewriteURLPath(req *http.Request) {
```
Comment (reasonerjt) [inline src/ui/api/chart_repository.go]: UT??

File: src/ui/api/chart_repository.go (hunk 3 above)
Comment (steven-zou) [reply src/ui/api/chart_repository.go]: UT is covered in Chart controller layer, will not provide UT cases in this layer

File: make/docker-compose.chartmuseum.tpl (hunk 8 above)
Comment (steven-zou) [reply make/docker-compose.chartmuseum.tpl]: We can remove

File: .travis.yml (hunk 7 above)
Comment (steven-zou) [reply .travis.yml]: Not used any more, removed

File: make/photon/chartserver/Dockerfile (hunk 9 above)
Comment (steven-zou) [reply make/photon/chartserver/Dockerfile]: sure

File: src/adminserver/systemcfg/systemcfg.go (hunk 11 above)
Comment (steven-zou) [reply src/adminserver/systemcfg/systemcfg.go]: Use consistent term chart_repositroy_url

**Pull request merged**